### DISCUSSION TITLE
Interpreter support

### DISCUSSION THREAD

**Discussion by notramo** [martenframework/marten#152] [Ideas]
Is it possible to support running the code in the interpreter in dev mode? Long compilation times are the biggest pain point for any Crystal app developmeent.

**Comment by ellmetha** [comment]
Hey! 👋 

Yes I agree that this would be awesome, and this has been in my mind since the introduction of the Crystal interpreter actually, but it's not possible to use Marten with the interpreter presently. The issue is that the Crystal interpreter is still pretty much unstable and experimental. All the attempts I made to use it on a Marten project have failed so far (and with the latest Crystal 1.11 version it's still not working with a barebone Marten project afaics).

We'll have to wait for it to be in a less experimental state I think.

**Reply by notramo** [reply]
What are the current problems? I tried with the stdlib `HTTP::Server` and a "hello world" Kemal project, and both worked. However, those are started with `crystal run src/main.cr`, and Marten uses its own CLI, so I could not test it.

https://github.com/crystal-lang/crystal/issues/12241 could be a blocker?

**Reply by ellmetha** [reply]
Maybe? But even just trying to interpret the entrypoint of a newly created project is going to break presently:

```bash
$ marten new project testproject
$ cd testproject
$ shards install
$ crystal i ./src/project.cr
Crystal interpreter 1.11.1 (2024-01-11).
EXPERIMENTAL SOFTWARE: if you find a bug, please consider opening an issue in
https://github.com/crystal-lang/crystal/issues/new/
Invalid memory access (signal 11) at address 0x(???)
[0x102c52470] *Exception::CallStack::print_backtrace:Nil +36 in /opt/homebrew/Cellar/crystal/1.11.1/bin/crystal
[0x102c5243c] ~procProc(Int32, Pointer(LibC::SiginfoT), Pointer(Void), Nil)@src/crystal/system/unix/signal.cr:131 +1672 in /opt/homebrew/Cellar/crystal/1.11.1/bin/crystal
libunwind: malformed __unwind_info at 0x102DCCF90 bad second level page
[0x1892a1a24] _sigtramp +56 in /usr/lib/system/libsystem_platform.dylib
```

And trying to interpret the server results in a seemingly hanging process:

```bash
$ crystal i ./src/server.cr
Crystal interpreter 1.11.1 (2024-01-11).
EXPERIMENTAL SOFTWARE: if you find a bug, please consider opening an issue in
https://github.com/crystal-lang/crystal/issues/new/

```

I haven't investigated this too much though. I don't consider this a priority for the framework presently, especially considering that the interpreter is still a very experimental feature at the moment.

**Reply by notramo** [reply]
On 1.11.2 the first memory error doesn't appear, but the server still hangs.
